### PR TITLE
Animation blend: add algorithm "freeform directional"

### PR DESCRIPTION
**Pull request by shrinktofit**
Re: #

### Changelog

*

-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

File: cocos/animation/marionette/blend-2d.ts [hunk 1]
```diff
@@ -256,50 +266,189 @@ const getGradientBandCartesianCoords: GetGradientBandCoords = (pI, pJ, input, pI
...
+            const direction = Vec2.copy(new Vec2(), example);
+            const magnitude = Vec2.len(direction);
+            exampleMagnitudes[iExample] = magnitude;
+            if (!approx(magnitude, 0.0, 1e-5)) {
+                Vec2.multiplyScalar(direction, direction, 1.0 / magnitude);
+            }
+            return direction;
+        });
+
+        const precomputedPIJs = this._precomputedPIJs = new Float32Array(PRECOMPUTED_VIJ_DATA_STRIDE * nExamples * nExamples);
```
Comment (SantyWang) [inline cocos/animation/marionette/blend-2d.ts]: What does `PIJ` means?

File: cocos/animation/marionette/blend-2d.ts [hunk 2]
```diff
@@ -256,50 +266,189 @@ const getGradientBandCartesianCoords: GetGradientBandCoords = (pI, pJ, input, pI
...
+                const directionJ = exampleDirections[jExample];
+                const averagedMagnitude = (magnitudeI + magnitudeJ) / 2;
+                const pOutput = PRECOMPUTED_VIJ_DATA_STRIDE * (nExamples * iExample + jExample);
+                precomputedPIJs[pOutput + 0] = (magnitudeJ - magnitudeI) / averagedMagnitude;
+                precomputedPIJs[pOutput + 1] = signedAngle(directionI, directionJ) * angleMultiplier;
+                precomputedPIJs[pOutput + 2] = averagedMagnitude;
+            }
+        }
+
+        this._cacheVIXAngles = new Float32Array(nExamples);
```
Comment (SantyWang) [inline cocos/animation/marionette/blend-2d.ts:323]: What does ‘VIX’ means?

File: cocos/animation/marionette/blend-2d.ts (hunk 1 above)
Comment (shrinktofit) [reply cocos/animation/marionette/blend-2d.ts]: I've documented it in https://github.com/cocos/cocos-engine/pull/14203/commits/325ebdfe0bc48e6f04be66f8ee062066296437c7 .

File: cocos/animation/marionette/blend-2d.ts (hunk 2 above)
Comment (shrinktofit) [reply cocos/animation/marionette/blend-2d.ts:323]: I've documented it in https://github.com/cocos/cocos-engine/pull/14203/commits/325ebdfe0bc48e6f04be66f8ee062066296437c7 .

**Review by SantyWang — APPROVED** [review]

**Review by SantyWang — APPROVED** [review]

**Pull request merged**